relay_fragment_sender_test.go

```diff
@@ -0,0 +1,272 @@
+
+	wbuf.WriteBytes([]byte("foo"))
+
+	tests := []struct {
+		msg                            string
+		frame                          *Frame
+		wantError                      string
+		sent                           bool
+		failure                        string
+		wantFailureRelayItemFuncCalled bool
+	}{
+		{
+			msg:   "successful send",
+			frame: f,
+			sent:  true,
+		},
+		{
+			msg:                            "send falure",
+			frame:                          f,
+			sent:                           false,
+			failure:                        "something bad happened",
+			wantFailureRelayItemFuncCalled: true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.msg, func(t *testing.T) {
+			wf := &writableFragment{
+				frame:    tt.frame,
+				contents: wbuf,
+			}
+
+			var failRelayItemFuncCalled bool
+
+			rfs := relayFragmentSender{
+				framePool: DefaultFramePool,
+				frameReceiver: &dummyFrameReceiver{
+					retSent:          tt.sent,
+					retFailureReason: tt.failure,
+				},
+				failRelayItemFunc: func(items *relayItems, id uint32, failure string) {
+					failRelayItemFuncCalled = true
+					assert.Equal(t, uint32(123), id, "got unexpected id")
+					assert.Equal(t, tt.failure, failure, "got unexpected failure string")
+				},
+				origID: 123,
+			}
+
+			err := rfs.flushFragment(wf)
+
+			if tt.wantError != "" {
+				require.EqualError(t, err, tt.wantError)
+				return
+			}
+			require.NoError(t, err)
+			assert.Equal(t, tt.wantFailureRelayItemFuncCalled, failRelayItemFuncCalled, "unexpected failRelayItemFunc called state")
+		})
+	}
+}
+
+type dummyArgWriter struct {
+	numCall      int
+	writeError   []string
+	closeError   string
+	bytesWritten []byte
+}
+
+func (w *dummyArgWriter) Write(b []byte) (int, error) {
+	retErr := w.writeError[w.numCall]
+	w.bytesWritten = append(w.bytesWritten, b...)
+	w.numCall++
+	if retErr != "" {
+		return 0, errors.New(retErr)
+	}
+	return len(b), nil
+}
+
+func (w *dummyArgWriter) Close() error {
+	if w.closeError != "" {
+		return errors.New(w.closeError)
+	}
+	return nil
+}
+
+func TestWriteArg2WithAppends(t *testing.T) {
+	tests := []struct {
+		msg             string
+		writer          *dummyArgWriter
+		arg2Map         map[string]string
+		overrideArg2Buf []byte
+		appends         []relay.KeyVal
+		wantError       string
+	}{
+		{
+			msg: "write success without appends",
+			writer: &dummyArgWriter{
+				writeError: []string{
+					"", // nh
+					"", // arg2
+				},
+			},
+			arg2Map: exampleArg2Map,
+		},
+		{
+			msg: "write success with appends",
+			writer: &dummyArgWriter{
+				writeError: []string{
+					"", // nh
+					"", // arg2
+					"", // key length
+					"", // key
+					"", // val length
+					"", // val
+				},
+			},
+			arg2Map: exampleArg2Map,
+			appends: []relay.KeyVal{
+				{Key: []byte("foo"), Val: []byte("bar")},
+			},
+		},
+		{
+			msg: "no nh in data",
+			writer: &dummyArgWriter{
+				writeError: []string{
+					"something went wrong", // nh
```

Comment:
suggest using `assert.AnError.String()` if you just want some error value for tests